src/imars3d/backend/workflow/engine.py

```diff
@@ -16,217 +16,207 @@ class WorkflowEngineError(RuntimeError):
-                    that are the outputs of other functions or are part of the metadata.
-            """
-            # load the ParameterizedFunction derived class associated to the function string
-            module_str = ".".join(function_str.split(".")[:-1])
-            module = importlib.import_module(module_str)
-            function_name = function_str.split(".")[-1]
-            f = getattr(module, function_name)
-
-            param_names = set(f.param.params().keys())
-            outputs = dict(function=f, globals_required=param_names.intersection(self.global_params))
-            return namedtuple("TaskFuncionInstrospection", outputs.keys())(**outputs)
-
-        def _update_registry(self, task: dict, function_outputs: Optional[Iterable[Any]] = None) -> None:
-            r"""
-            Fetch the values for the global parameters that have just been set or computed.
-
-            Parameters
-            ----------
-            task
-                task entry as it appears in the JSON configuration file
-            function_outputs
-                return value of the function carrying out the task. Should be an iterable, such as a ``List``.
-
-            Raises
-            ------
-            WorkflowEngineError
-                The outputs, as defined in the task entry, don't match the actual return value of the function.
-            """
-            # insert explicitly passed globals values
+        errors = " and ".join([k for k, v in tests.items() if v is False])
+        if errors:
+            raise WorkflowEngineError(errors)
+
+    def __init__(self, global_params: Union[Set[str], List[str]]) -> None:
+        r"""
+
+        Parameters
+        ----------
+        global_params
+            List of JSON entries deemed as accessible by any workflow task. These are the metadata
+            and the task outputs.
+        """
+        self._global_params: set = set(global_params)
+        self._registry: Optional[dict] = None  # will store set or computed globals parameters
+
+    def _instrospect_task_function(self, function_str: str) -> namedtuple:
+        r"""Obtain information from the function associated to one task in the workflow.
+
+        Parameters
+        ----------
+        function_str
+            Literal name for the ``ParameterizedFunction` derived class
+
+        Returns
+        -------
+        namedtuple
+            Fields of the namedtuple ``TaskFuncionInstrospection``:
+            function : ParameterizedFunction
+                a reference to the function object
+            globals_required : set[str]
+                the function arguments that are globals, i.e., the names of param.Parameter objects
+                that are the outputs of other functions or are part of the metadata.
+        """
+        # load the ParameterizedFunction derived class associated to the function string
+        module_str = ".".join(function_str.split(".")[:-1])
```

Comment:
Since this is duplicated in [`validate._function_parts`](https://github.com/ornlneutronimaging/iMars3D/blob/next/src/imars3d/backend/workflow/validate.py#L80-L84) would it be good to split it off into a static function that can be reused? I realize it is only 2 lines of code, but my experience is that introspection is pretty brittle between versions of python.

Reply:
Renamed `validate._function_parts` into `validate.function_parts` to "make it public" and now use it in `_instrospect_task_function`